winpr/libwinpr/sspi/Negotiate/negotiate.c

```diff
@@ -305,32 +872,263 @@ static SECURITY_STATUS SEC_ENTRY negotiate_AcceptSecurityContext(
     ULONG TargetDataRep, PCtxtHandle phNewContext, PSecBufferDesc pOutput, PULONG pfContextAttr,
     PTimeStamp ptsTimeStamp)
 {
+	NEGOTIATE_CONTEXT* context = NULL;
+	NEGOTIATE_CONTEXT init_context = { 0 };
+	MechCred* creds;
+	PCredHandle sub_cred = NULL;
+	NegToken input_token = { NOSTATE, 0 };
+	NegToken output_token = { NOSTATE, 0 };
+	PSecBuffer input_buffer = NULL;
+	PSecBuffer output_buffer = NULL;
+	SecBufferDesc mech_input = { SECBUFFER_VERSION, 1, &input_token.mechToken };
+	SecBufferDesc mech_output = { SECBUFFER_VERSION, 1, &output_token.mechToken };
 	SECURITY_STATUS status;
-	NEGOTIATE_CONTEXT* context;
-	context = (NEGOTIATE_CONTEXT*)sspi_SecureHandleGetLowerPointer(phContext);
+	BYTE *p, tag;
+	ULONG bytes_remain, len;
+	sspi_gss_OID_desc oid = { 0 };
+
+	if (!phCredential || !SecIsValidHandle(phCredential))
+		return SEC_E_NO_CREDENTIALS;
+
+	if (!pInput)
+		return SEC_E_INVALID_TOKEN;
+
+	context = sspi_SecureHandleGetLowerPointer(phContext);
+	creds = sspi_SecureHandleGetLowerPointer(phCredential);
+	input_buffer = sspi_FindSecBuffer(pInput, SECBUFFER_TOKEN);
+	if (pOutput)
+		output_buffer = sspi_FindSecBuffer(pOutput, SECBUFFER_TOKEN);
 
 	if (!context)
 	{
-		context = negotiate_ContextNew();
+		/* Check for NTLM token */
+		if (strncmp(input_buffer->pvBuffer, "NTLMSSP", 8) == 0)
```

Comment:
We shall check that `input_buffer->cbBuffer` is at least 8 before doing this comparison